### PR TITLE
Show logical structure for Collection and Map instances

### PR DESCRIPTION
**Pull request by testforstephen**
Signed-off-by: Jinbo Wang <jinbwan@microsoft.com>

image

Close Microsoft/vscode-java-debug#227

### CODE REVIEW

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/VariablesRequestHandler.java [hunk 1]
```diff
@@ -109,10 +116,41 @@
         } else {
             try {
                 ObjectReference containerObj = (ObjectReference) containerNode.getProxiedVariable();
-                if (varArgs.count > 0) {
-                    childrenList = VariableUtils.listFieldVariables(containerObj, varArgs.start, varArgs.count);
-                } else {
-                    childrenList = VariableUtils.listFieldVariables(containerObj, showStaticVariables);
+                if (DebugSettings.getCurrent().showLogicalStructure) {
```
Comment (andxu) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/VariablesRequestHandler.java]: wrap this if in a new try catch and fail back to the non-logical one logic.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/IEvaluationProvider.java [hunk 2]
```diff
@@ -41,6 +42,17 @@
      */
     CompletableFuture<Value> evaluate(String expression, ThreadReference thread, int depth);
 
+    /**
+     * Evaluate the expression at the given context of the object instance, return the promise which is to be resolved/rejected when
+     * the evaluation finishes.
+     * @param expression The expression to be evaluated
+     * @param context The jdi object which the expression will be executed at
```
Comment (andxu) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/IEvaluationProvider.java]: do you mean thisObj?

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/IEvaluationProvider.java [hunk 3]
```diff
@@ -41,6 +42,17 @@
...
+     * @param thread The jdi thread which the expression will be executed at
+     * @param depth The depth of stackframe of the stopped thread
+     * @return the evaluation result future
+     */
+    CompletableFuture<Value> evaluate(String expression, ObjectReference context, ThreadReference thread, int depth);
```
Comment (andxu) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/IEvaluationProvider.java]: rename this function to evaluateOnObject, two functions with same name and trivial differences in parameters is not good, caller will get confused.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/variables/VariableProxy.java [hunk 4]
```diff
@@ -30,10 +31,11 @@
      * @param variable
      *              the variable object
      */
-    public VariableProxy(ThreadReference thread, String scopeName, Object variable) {
+    public VariableProxy(ThreadReference thread, String scopeName, Object variable, StackFrameReference stackFrame) {
         this.thread = thread;
         this.scopeName = scopeName;
         this.variable = variable;
+        this.stackFrame = stackFrame;
         hashCode = thread.hashCode() & scopeName.hashCode() & variable.hashCode();
```
Comment (andxu) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/variables/VariableProxy.java]: hasCode should be changed too

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/variables/VariableProxy.java [hunk 5]
```diff
@@ -30,10 +31,11 @@
      * @param variable
      *              the variable object
      */
```
Comment (andxu) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/variables/VariableProxy.java]: I find you only use the depth of the stackFrame, use a depth instead?

File: com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/eval/JdtEvaluationProvider.java [hunk 6]
```diff
@@ -108,6 +112,28 @@ public void initialize(IDebugAdapterContext context, Map<String, Object> props)
...
+            ensureDebugTarget(thread.virtualMachine(), thread, depth);
+            JDIThread jdiThread = getMockJDIThread(thread);
+            JDIStackFrame stackframe = createStackFrame(jdiThread, depth);
+            if (stackframe == null) {
+                throw new IllegalStateException("Cannot evaluate because the stackframe is not available.");
+            }
+
+            JDIObjectValue jdiObject = new JDIObjectValue(debugTarget, context);
+            ASTEvaluationEngine engine = new ASTEvaluationEngine(project, debugTarget);
+            ICompiledExpression compiledExpression = engine.getCompiledExpression(expression, jdiObject);
```
Comment (andxu) [inline com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/eval/JdtEvaluationProvider.java]: for conditional breakpoints, the compiled expression is saved into breakpoint for better performance, convert evaluation text to compiled expression is heavy, be aware of that and take stress tests on your code.

Comment (testforstephen) [reply com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/eval/JdtEvaluationProvider.java]: The variable evaluation is requested when the user click the complex object to expand its children. And a single click only triggers one or two evaluation action. No need of stress test.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/variables/VariableProxy.java (hunk 5 above)
Comment (testforstephen) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/variables/VariableProxy.java]: nice.

**Pull request merged**